### PR TITLE
[FLINK-23969][connector/pulsar] Create e2e tests for pulsar connector.

### PR DESCRIPTION
**Pull request by syhily**
## What is the purpose of the change

Write a test application using Pulsar Source and execute it in distributed fashion. Check fault-tolerance by crashing and restarting a TM.

Ideally, we test different subscription modes and sticky keys in particular.

## Brief change log

* Change the testutils in pulsar connector for using in container.
* Create pulsar e2e tests which supports four types of subscription.

## Verifying this change

This change added tests and can be verified as follows:

  - *Added integration tests for end-to-end tests from pulsar to flink.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)
  - If yes, how is the feature documented? (**not applicable** / docs / JavaDocs / not documented)

### CODE REVIEW

File: flink-end-to-end-tests/flink-end-to-end-tests-pulsar/src/test/java/org/apache/flink/tests/util/pulsar/PulsarSourceOrderedE2ECase.java
Comment (fapaul) [inline flink-end-to-end-tests/flink-end-to-end-tests-pulsar/src/test/java/org/apache/flink/tests/util/pulsar/PulsarSourceOrderedE2ECase.java:39]: The `testIdleReader` fails consistenly on my local machine with the following error message

```
Caused by: java.util.concurrent.CompletionException: org.apache.pulsar.client.admin.internal.http.AsyncHttpConnector$RetryException: Could not complete the operation. Number of retries has been exhausted. Failed reason: java.lang.OutOfMemoryError: Direct buffer memory
	at java.base/java.util.concurrent.CompletableFuture.encodeThrowable(CompletableFuture.java:331)
	at java.base/java.util.concurrent.CompletableFuture.completeThrowable(CompletableFuture.java:346)
	at java.base/java.util.concurrent.CompletableFuture$OrApply.tryFire(CompletableFuture.java:1503)
	... 42 more
Caused by: org.apache.pulsar.client.admin.internal.http.AsyncHttpConnector$RetryException: Could not complete the operation. Number of retries has been exhausted. Failed reason: java.lang.OutOfMemoryError: Direct buffer memory
	at org.apache.pulsar.client.admin.internal.http.AsyncHttpConnector.lambda$retryOperation$4(AsyncHttpConnector.java:249)
	... 39 more
Caused by: org.apache.pulsar.shade.io.netty.handler.codec.EncoderException: java.lang.OutOfMemoryError: Direct buffer memory
	at org.apache.pulsar.shade.io.netty.handler.codec.MessageToMessageEncoder.write(MessageToMessageEncoder.java:104)
	at org.apache.pulsar.shade.io.netty.channel.CombinedChannelDuplexHandler.write(CombinedChannelDuplexHandler.java:346)
	at org.apache.pulsar.shade.io.netty.channel.AbstractChannelHandlerContext.invokeWrite0(AbstractChannelHandlerContext.java:717)
	at org.apache.pulsar.shade.io.netty.channel.AbstractChannelHandlerContext.invokeWrite(AbstractChannelHandlerContext.java:709)
	at org.apache.pulsar.shade.io.netty.channel.AbstractChannelHandlerContext.write(AbstractChannelHandlerContext.java:792)
	at org.apache.pulsar.shade.io.netty.channel.AbstractChannelHandlerContext.write(AbstractChannelHandlerContext.java:702)
	at org.apache.pulsar.shade.io.netty.handler.stream.ChunkedWriteHandler.doFlush(ChunkedWriteHandler.java:303)
	at org.apache.pulsar.shade.io.netty.handler.stream.ChunkedWriteHandler.flush(ChunkedWriteHandler.java:132)
	at org.apache.pulsar.shade.io.netty.channel.AbstractChannelHandlerContext.invokeFlush0(AbstractChannelHandlerContext.java:750)
	at org.apache.pulsar.shade.io.netty.channel.AbstractChannelHandlerContext.invokeWriteAndFlush(AbstractChannelHandlerContext.java:765)
	at org.apache.pulsar.shade.io.netty.channel.AbstractChannelHandlerContext.write(AbstractChannelHandlerContext.java:790)
	at org.apache.pulsar.shade.io.netty.channel.AbstractChannelHandlerContext.writeAndFlush(AbstractChannelHandlerContext.java:758)
	at org.apache.pulsar.shade.io.netty.channel.DefaultChannelPipeline.writeAndFlush(DefaultChannelPipeline.java:1020)
	at org.apache.pulsar.shade.io.netty.channel.AbstractChannel.writeAndFlush(AbstractChannel.java:311)
	at org.apache.pulsar.shade.org.asynchttpclient.netty.request.NettyRequestSender.writeRequest(NettyRequestSender.java:420)
	... 23 more
```

It seems that all subsequent tests stall because the pulsar broker seems to be down

```
8415 [pulsar-client-io-1-1] WARN  org.apache.pulsar.client.impl.ConnectionHandler [] - [persistent://public/default/pulsar-multiple-topic-0-ESYvbdZk-partition-0] [standalone-0-32] Could not get connection to broker: org.apache.pulsar.client.api.PulsarClientException: java.util.concurrent.CompletionException: org.apache.pulsar.shade.io.netty.channel.AbstractChannel$AnnotatedConnectException: Connection refused: localhost/127.0.0.1:52267 -- Will try again in 46.344 s
```

**Review by fapaul — CHANGES_REQUESTED** [review]
Thanks for adding the e2e tests they all look reasonable.

I'd prefer having one single class executing all tests if possible IMO the separated classes are only other test cases that do not require separate setups.

This way it is probably also possible to stick with the existing annotations and use the `@TestTemplate` instead of `@Test` similar to the tests implemented in `SourceTestSuiteBase`.

File: flink-connectors/flink-connector-pulsar/src/main/java/org/apache/flink/connector/pulsar/source/enumerator/PulsarSourceEnumerator.java
Comment (syhily) [reply flink-connectors/flink-connector-pulsar/src/main/java/org/apache/flink/connector/pulsar/source/enumerator/PulsarSourceEnumerator.java:213]: This is a workaround for the current pulsar release.
The related issue is described in https://github.com/apache/pulsar/pull/12035 and https://issues.apache.org/jira/browse/FLINK-24283

File: flink-end-to-end-tests/flink-end-to-end-tests-pulsar/src/test/java/org/apache/flink/tests/util/pulsar/FlinkContainerWithPulsarEnvironment.java
Comment (syhily) [reply flink-end-to-end-tests/flink-end-to-end-tests-pulsar/src/test/java/org/apache/flink/tests/util/pulsar/FlinkContainerWithPulsarEnvironment.java:30]: I'm quite familiar with `KafkaSourceE2ECase`. But its test logic doesn't apply to pulsar's `Shared` and `Key_Shared` subscriptions. The `Shared` and `Key_Shared` subscriptions would share the topic partition for all the readers. This leads to the messages would be consumed in an unordered way. So we can not use `SourceTestSuiteBase`.

File: flink-end-to-end-tests/flink-end-to-end-tests-pulsar/src/test/java/org/apache/flink/tests/util/pulsar/PulsarSourceKeySharedE2ECase.java
Comment (syhily) [reply flink-end-to-end-tests/flink-end-to-end-tests-pulsar/src/test/java/org/apache/flink/tests/util/pulsar/PulsarSourceKeySharedE2ECase.java:58]: It's used for Sonarlint in IDEA. Since it's a test code, I think we can just keep it.

File: flink-end-to-end-tests/flink-end-to-end-tests-pulsar/src/test/java/org/apache/flink/tests/util/pulsar/PulsarSourceKeySharedE2ECase.java
Comment (syhily) [reply flink-end-to-end-tests/flink-end-to-end-tests-pulsar/src/test/java/org/apache/flink/tests/util/pulsar/PulsarSourceKeySharedE2ECase.java:97]: Nope, the zero is accepted. The hash range is between 0 and 65535.

File: flink-end-to-end-tests/flink-end-to-end-tests-pulsar/src/test/java/org/apache/flink/tests/util/pulsar/PulsarSourceOrderedE2ECase.java
Comment (syhily) [reply flink-end-to-end-tests/flink-end-to-end-tests-pulsar/src/test/java/org/apache/flink/tests/util/pulsar/PulsarSourceOrderedE2ECase.java:39]: Pulsar's docker image consumes quite a large amount of memory. It would start a pulsar standalone, a bookkeeper, and a zookeeper. I think you may need to increase the docker's memory on your local machine.

This issue is not related to the test code. You can see that all the tests passed on flink ci.

File: flink-end-to-end-tests/flink-end-to-end-tests-pulsar/src/test/java/org/apache/flink/tests/util/pulsar/FlinkContainerWithPulsarEnvironment.java
Comment (fapaul) [reply flink-end-to-end-tests/flink-end-to-end-tests-pulsar/src/test/java/org/apache/flink/tests/util/pulsar/FlinkContainerWithPulsarEnvironment.java:30]: One thing you could is the following for both test classes `PulsarSourceKeySharedE2ECase` `PulsarSourceSharedE2ECase` 

```java
@ExtendWith({
        ConnectorTestingExtension.class,
        TestLoggerExtension.class,
        TestCaseInvocationContextProvider.class
})
@TestInstance(TestInstance.Lifecycle.PER_CLASS)
@SuppressWarnings("java:S3577")
class PulsarSourceKeySharedE2ECase {

    /** Flink runtime. */
    @TestEnv
    final FlinkContainerWithPulsarEnvironment flink = new FlinkContainerWithPulsarEnvironment(1, 6);

    /** Pulsar runtime. */
    @ExternalSystem
    final PulsarTestEnvironment pulsar =
            new PulsarTestEnvironment(container(flink.getFlinkContainer()));

    @TestTemplate
    @DisplayName(....)
    void testConsumingMessageFromTheSpecifiedHashRange(TestEnvironment testEnv, ExternalContext<T> externalContext) throws Exception {
....
    }
}
```
 
I think it makes sense to reuse the existing annotations here and you can get rid of the additional `BeforeAllCallback` and `AfterAllCallback`.

Comment (syhily) [reply flink-end-to-end-tests/flink-end-to-end-tests-pulsar/src/test/java/org/apache/flink/tests/util/pulsar/FlinkContainerWithPulsarEnvironment.java:30]: @fapaul Yep, you are right. I'll refactor the code by following your advice.

File: flink-end-to-end-tests/flink-end-to-end-tests-pulsar/src/test/java/org/apache/flink/tests/util/pulsar/PulsarSourceOrderedE2ECase.java
Comment (fapaul) [reply flink-end-to-end-tests/flink-end-to-end-tests-pulsar/src/test/java/org/apache/flink/tests/util/pulsar/PulsarSourceOrderedE2ECase.java:39]: Hmm after increasing the available memory to 8G it kept happening with Java 11. Once I ran the tests with Java 8 everything seemed to be fine. Can you double-check whether the e2e tests run on Java 11? AFAIK the CI only runs with java 11 nightly and not on PRs

Comment (syhily) [reply flink-end-to-end-tests/flink-end-to-end-tests-pulsar/src/test/java/org/apache/flink/tests/util/pulsar/PulsarSourceOrderedE2ECase.java:39]: Oh, I see. I think the reason is that under Java 11, Netty will allocate memory from the pool of Java Direct Memory and is affected by the MaxDirectMemory limit. Under Java 8, it allocates native memory and is not affected by that setting. We may need time to fix this error on the Pulsar side, there is [an issue](https://github.com/apache/pulsar/issues/9721) about this.

Since Flink is still running under Java 8, can we skip these tests under Java 11?

Comment (fapaul) [reply flink-end-to-end-tests/flink-end-to-end-tests-pulsar/src/test/java/org/apache/flink/tests/util/pulsar/PulsarSourceOrderedE2ECase.java:39]: Hmm if we want to disable it we probably have to do it via maven by checking which java profile is active https://github.com/apache/flink/blob/f74db4b6cce9a28f484dc9d5e6470a50b15c121a/flink-end-to-end-tests/run-nightly-tests.sh#L264 

Is this issue also affecting the production code?

Comment (syhily) [reply flink-end-to-end-tests/flink-end-to-end-tests-pulsar/src/test/java/org/apache/flink/tests/util/pulsar/PulsarSourceOrderedE2ECase.java:39]: We won't meet this issue in production code unless we use Java 11.

Comment (syhily) [reply flink-end-to-end-tests/flink-end-to-end-tests-pulsar/src/test/java/org/apache/flink/tests/util/pulsar/PulsarSourceOrderedE2ECase.java:37]: I resolve this issue and passed on my local machine with JDK 11.

**Comment by syhily** [comment]
@flinkbot run azure

**Comment by syhily** [comment]
@flinkbot run azure

File: flink-end-to-end-tests/flink-end-to-end-tests-pulsar/src/test/java/org/apache/flink/tests/util/pulsar/common/UnorderedSourceTestSuiteBase.java
Comment (fapaul) [inline flink-end-to-end-tests/flink-end-to-end-tests-pulsar/src/test/java/org/apache/flink/tests/util/pulsar/common/UnorderedSourceTestSuiteBase.java:51]: Nit: I am not sure using inheritance here is strictly necessary currently there is only one class inheriting from it.

File: flink-end-to-end-tests/flink-end-to-end-tests-pulsar/src/test/java/org/apache/flink/tests/util/pulsar/common/FlinkContainerWithPulsarEnvironment.java
Comment (fapaul) [inline flink-end-to-end-tests/flink-end-to-end-tests-pulsar/src/test/java/org/apache/flink/tests/util/pulsar/common/FlinkContainerWithPulsarEnvironment.java:50]: Did you already open a ticket to fix this properly? I think we also need to update the release notes to mention the pulsar source is for now only supporting Java 8. 

FYI @dawidwys

**Review by fapaul — APPROVED** [review]
Thanks for addressing all the comments. I ran all the tests locally with java 8 and 11 and all seem to pass now 👍 

I left one nitty suggestion feel free to ignore it but we should definitely mention in the release notes that the Pulsar source is currently only safe to use with java 8.

File: flink-end-to-end-tests/flink-end-to-end-tests-pulsar/src/test/java/org/apache/flink/tests/util/pulsar/common/UnorderedSourceTestSuiteBase.java
Comment (syhily) [reply flink-end-to-end-tests/flink-end-to-end-tests-pulsar/src/test/java/org/apache/flink/tests/util/pulsar/common/UnorderedSourceTestSuiteBase.java:51]: I will add other e2e tests for Pulsar schema later in Flink 1.15 release. Let's just keep this inheritance here.

File: flink-end-to-end-tests/flink-end-to-end-tests-pulsar/src/test/java/org/apache/flink/tests/util/pulsar/common/FlinkContainerWithPulsarEnvironment.java
Comment (syhily) [reply flink-end-to-end-tests/flink-end-to-end-tests-pulsar/src/test/java/org/apache/flink/tests/util/pulsar/common/FlinkContainerWithPulsarEnvironment.java:50]: @fapaul The related issue in Pulsar is [this](https://github.com/apache/pulsar/issues/9721). Pulsar would add a global (static) memory limit in **PIP-74**: https://github.com/apache/pulsar/wiki/PIP-74%3A-Pulsar-client-memory-limits.

Comment (fapaul) [reply flink-end-to-end-tests/flink-end-to-end-tests-pulsar/src/test/java/org/apache/flink/tests/util/pulsar/common/FlinkContainerWithPulsarEnvironment.java:50]: @syhily Can you also create a FLINK ticket to track the progress of this effort? Eventually, we have to increase the pulsar version once this is merged to support Java 11.

The Flink ticket we can then reference in the release notes.

Comment (syhily) [reply flink-end-to-end-tests/flink-end-to-end-tests-pulsar/src/test/java/org/apache/flink/tests/util/pulsar/common/FlinkContainerWithPulsarEnvironment.java:50]: @fapaul I have created a JIRA for tracking this issue. https://issues.apache.org/jira/projects/FLINK/issues/FLINK-24233

**Comment by syhily** [comment]
It seems that all the E2E tests were passed on Java 8 and 11. I think that the code condition matches the review requirements. Can we merge this into Flink 1.14.0 release? The direct memory leak on Java 11 has an extra JIRA. We can just wait for the new Pulsar release.

**Comment by fapaul** [comment]
@syhily can you also create a PR for the backport to the `release-1.14` branch?

**Comment by syhily** [comment]
> @syhily can you also create a PR for the backport to the `release-1.14` branch?

OK, the backport [PR](https://github.com/apache/flink/pull/17304) has been created.

**Pull request merged**
